lib/recurly/schema.rb

```diff
@@ -46,71 +45,108 @@ def self.get_recurly_class(type)
+      # @return [Class,Symbol]
+      attr_reader :type
+
+      PRIMITIVE_TYPES = [
+        String,
+        Integer,
+        Float,
+        Hash,
+      ].freeze
+
+      def self.build(type, options = {})
+        if PRIMITIVE_TYPES.include? type
+          PrimitiveAttribute.new(type)
+        elsif type == :Boolean
+          BooleanAttribute.new
+        elsif type == DateTime
+          DateTimeAttribute.new
+        elsif type.is_a? Symbol
+          ResourceAttribute.new(type)
+        elsif type == Array
+          item_attr = build(options[:item_type])
+          ArrayAttribute.new(item_attr)
+        else
+          throw ArgumentError
+        end
+      end
 
-      def initialize(item_type)
-        @item_type = item_type
+      def initialize(type = nil)
+        @type = type
       end
 
-      def to_s
-        "List<#{item_type}>"
+      def cast(value)
+        value
       end
-    end
 
-    # Describes and attribute for a schema.
-    class Attribute
-      # The name of the attribute.
-      # @return [Symbol]
-      attr_accessor :name
+      def recurly_class
+        @recurly_class ||= Schema.get_recurly_class(type)
+      end
+    end
 
-      # The type of the attribute. Might be a class like `DateTime`
-      # or could be a Recurly object. In this case a symbol should be used.
-      # Example: :Account
-      # @return [Class,Symbol]
-      attr_accessor :type
+    class PrimitiveAttribute < Attribute
+      def is_valid?(value)
+        value.is_a? self.type
+      end
+    end
 
-      # Options for the attribute.
-      # @return [Hash]
-      attr_accessor :options
+    class BooleanAttribute < Attribute
+      def is_valid?(value)
+        [true, false].include? value
+      end
+    end
 
-      # The description of the attribute for documentation.
-      # @return [String]
-      attr_accessor :description
+    class DateTimeAttribute < Attribute
+      def is_valid?(value)
+        value.is_a?(String) || value.is_a?(DateTime)
+      end
 
-      def initialize(name, type, options = {}, description = nil)
-        @name = name
-        @type = type
-        @options = options
-        @description = description
+      def cast(value)
+        if value.is_a?(DateTime)
+          value
+        else
+          DateTime.parse(value)
+        end
       end
 
-      def read_only?
-        @options.fetch(:read_only, false)
+      def type
+        DateTime
       end
+    end
 
-      def recurly_class
-        Schema.get_recurly_class(type == Array ? options[:item_type] : type)
+    class ResourceAttribute < Attribute
+      def is_valid?(value)
+        value.is_a? Hash
       end
 
-      def is_primitive?
-        t = type == Array ? options[:item_type] : type
-        t.is_a?(Class) || t == :Boolean
+      def cast(value)
+        self.recurly_class.cast(value)
       end
     end
 
-    private_constant :List
-    private_constant :Attribute
+    class ArrayAttribute < Attribute
```

Comment:
Each Attribute type now has it's own class which contains the logic for validating and casting. This means we don't have to run through conditionals to find the right code to validate or cast a value. it's reduced to a hash lookup.

Reply:
These are also easier to test or change in isolation.